### PR TITLE
Revisit RootErrorBoundary

### PR DESCRIPTION
**Pull request by brentvatne**
# Why

https://github.com/expo/expo/issues/4581
https://github.com/expo/expo/issues/4581

The purpose of RootErrorBoundary is to catch errors that happen while the splash screen is visible and then hide the splash screen, because it is confusing to people to see the app frozen on the splash screen.

On iOS this happens already automatically. On Android, we need to render some content to hide the splash screen.

# How

- Add `getDerivedStateFromError` so we stop getting a warning about it.
- Rethrow error from `componentDidCatch` - it doesn't hurt to do this but I could not observe any differences in the behaviour as described in another thread.
- Remove `RootErrorBoundary` on web - it's not needed, and is covered up quickly by the Webpack error screen.
- Remove `RootErrorBoundary` on iOS - it's not needed, it's handled automatically.
- Add notes about the implementation in comments.
- Removed the special UI that tells people about the root component failing to render and just render null instead. This is not particularly useful and users should instead refer to the red screen error. We may want to consider looking into making it not possible to hide the red screen if the app has failed to start.

# Follow-up

- Cherry-pick to sdk-37 branch and release.
- We should see if we can get rid of this entirely and make `AppLoading` and `SplashScreen` hide themselves on fatal errors.
- If we can't do that, then we should integrate RootErrorBoundary with SplashScreen.
- Should red screen even be dismissible if the splash screen was still active while a fatal error occurred? It's not helpful to see a blank white screen when you dismiss.

# Test Plan

Run this app and verify that the splash screen is hidden on Android and iOS and that you see the appropriate webpack error on web:

```jsx
export default function App() {
  let [isReady, setIsReady] = React.useState(false);

  React.useEffect(() => {
    setTimeout(() => {
      throw new Error("wat");
    }, 1000);
  }, []);

  if (!isReady) {
    return <AppLoading />;
  }

  return (
    <View style={styles.container}>
      <Text>Open up App.js to start working on your app!</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center",
  },
});
```

### CODE REVIEW

File: packages/expo/src/launch/withExpoRoot.web.tsx [hunk 1]
```diff
@@ -14,15 +13,7 @@ export default function withExpoRoot<P extends InitialProps>(
         exp: { ...this.props.exp, errorRecovery: ErrorRecovery.recoveredProps },
```
Comment (EvanBacon) [inline packages/expo/src/launch/withExpoRoot.web.tsx:13]: I'm not sure if we still need these ErrorRecovery props, not sure what they do or where they're used.

Comment (brentvatne) [reply packages/expo/src/launch/withExpoRoot.web.tsx:13]: from this thing https://docs.expo.io/versions/latest/sdk/error-recovery/

> If your project crashes in production as a result of a fatal JS error, Expo will reload your project. If you've set these props, they'll be passed to your reloaded project's initial props under exp.errorRecovery

**Comment by ide** [comment]
I think it would be quite nice to have tests for this code, especially because (a) testing it currently involves throwing an error manually and (b) behavior differs across development and production, which we could mock out via `__DEV__` in a unit test. ReactTestRenderer has `act()` which can run hooks. I am not sure how it works with componentDidCatch and getDerivedStateFromError but we could give it a shot.

File: packages/expo/src/launch/__tests__/withExpoRoot-test.web.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import * as React from 'react';
+import renderer, { ReactTestInstance } from 'react-test-renderer';
+
+import AppLoading from '../AppLoading';
+import RootErrorBoundary from '../RootErrorBoundary';
+import withExpoRoot from '../withExpoRoot';
+import { withDEV } from './utils';
+
+it('does RootErrorBoundary in DEV=true or DEV=false', () => {
```
Comment (ide) [inline packages/expo/src/launch/__tests__/withExpoRoot-test.web.tsx:9]:
```suggestion
it('does not add RootErrorBoundary in DEV=true or DEV=false', () => {
```
?

File: packages/expo/src/launch/__tests__/withExpoRoot-test.ios.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import * as React from 'react';
+import renderer, { ReactTestInstance } from 'react-test-renderer';
+
+import AppLoading from '../AppLoading';
+import RootErrorBoundary from '../RootErrorBoundary';
+import withExpoRoot from '../withExpoRoot';
+import { withDEV } from './utils';
+
+it('does RootErrorBoundary in DEV=true or DEV=false', () => {
```
Comment (ide) [inline packages/expo/src/launch/__tests__/withExpoRoot-test.ios.tsx:9]:
```suggestion
it('does not add RootErrorBoundary in DEV=true or DEV=false', () => {
```

**Review by ide — APPROVED** [review]
There's a lint error but other than that this looks good. Added some inline comments for the test descriptions.

File: packages/expo-error-recovery/src/ErrorRecovery.fx.ts [hunk 4]
```diff
@@ -14,7 +15,9 @@ if (Platform.OS !== 'web') {
     globalHandler(error, isFatal);
   });
 } else {
-  window.addEventListener('error', () => {
-    ExpoErrorRecovery.saveRecoveryProps(getRecoveryPropsToSave());
-  });
+  if (canUseDOM) {
```
Comment (lukmccall) [inline packages/expo-error-recovery/src/ErrorRecovery.fx.ts:18]: I think that `else if` will look better ;)

**Review by lukmccall — APPROVED** [review]
LGTM 🏅
And don't forget to rebuild the `expo-error-recovery` package.

File: packages/expo/src/launch/__tests__/utils.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    console.error = () => {};
+    fn();
+  } catch (error) {
+    return error;
+  } finally {
+    console.error = consoleError;
+  }
+}
+
+export function withDEV(dev: boolean, fn: any) {
```
Comment (bbarthec) [inline packages/expo/src/launch/__tests__/utils.ts:13]:
```suggestion
export function withDEV(dev: boolean, fn: () => void) { // or Function?
```

**Review by bbarthec — APPROVED** [review]

**Pull request merged**